Fix problem with string data allocation TG-1612

romainbrenguier opened the pull request:
String data was allocated on a stack as an infinite array of characters. This could cause problems when this happens in a loop or a function call as several strings could have the same data pointer. We fix that by using dynamic allocation instead.

romainbrenguier committed:
Fix allocation of infinite char arrays

Infinite char arrays used in string need to be dynamically allocated
instead of being on the stack as it used to be.

romainbrenguier committed:
Add invariant on array to pointer association

In the string solver each char array should be associated to a pointer
which is unique for each array. This invariant ensures we don't make
mistake there.

romainbrenguier committed:
Adapt unit test for allocation of string data

allredj commented on src/java_bytecode/java_object_factory.cpp:
```diff
@@ -527,15 +527,19 @@ static mp_integer max_value(const typet &type)
-///   .data=nondet_infinite_array$2 };
+/// *string_data_pointer = nondet_infinite_array$2;
+/// cprover_associate_array_to_pointer_func(
+///   *string_data_pointer, *string_data_pointer);
```
Why is it not `string_data_pointer, *string_data_pointer);`

romainbrenguier replied:
It is actually `*string_data_pointer` and `&(*string_data_pointer[0])`, the first is the array and the second the pointer. Both are printed the same by CBMC, and I just did a copy past of the goto program output.

allredj commented on src/java_bytecode/java_object_factory.cpp:
```diff
@@ -527,15 +527,19 @@ static mp_integer max_value(const typet &type)
+/// cprover_associate_array_to_pointer_func(
+///   *string_data_pointer, *string_data_pointer);
+/// cprover_associate_length_to_array_func(
+///   *string_data_pointer, tmp_object_factory);
```
Why is it not `string_data_pointer, tmp_object_factory);`

romainbrenguier replied:
string_data_pointer is a pointer to the data (which is the array)

allredj commented on src/java_bytecode/java_object_factory.cpp:
```diff
@@ -593,21 +598,47 @@ codet initialize_nondet_string_struct(
+    side_effect_exprt java_new_array(ID_allocate);
+    java_new_array.copy_to_operands(infinity_exprt(java_int_type()));
+    java_new_array.copy_to_operands(false_exprt());
+    java_new_array.type() = data_pointer.type();
```
We should have/use a constructor instead of assigning fields in that way. (maybe add a todo).

LAJW replied:
Or just pop off into another static `make_` function. Adding another constructor might not be necessary if this is domain specific.

romainbrenguier replied:
yes I constructor is maybe not the best way as there is no `allocate_exprt`, a `make_allocate_expr` could be useful if it is used in several places.

allredj reviewed (commented):
Looks good. Need to investigate failing test on test-gen.

LAJW commented on src/java_bytecode/java_object_factory.cpp:
```diff
@@ -593,21 +598,47 @@ codet initialize_nondet_string_struct(
+    const dereference_exprt data_expr(data_pointer);
+    const exprt nondet_array =
+      make_nondet_infinite_char_array(symbol_table, loc, code);
+    code.add(code_assignt(data_expr, nondet_array));
```
Suggestion: Each of those sections starting with a comment, and ending with a `code.add` could be put into a dedicated function named after the comment. For instance here - `make_assign_nondet_infinite_char_array` or something similar. That way, you could express the intent of a code section with the code itself.

LAJW commented on src/solvers/refinement/string_constraint_generator_main.cpp:
```diff
@@ -270,7 +270,10 @@ exprt string_constraint_generatort::associate_array_to_pointer(
   /// \todo We should use a function for inserting the correspondance
   /// between array and pointers.
-  arrays_of_pointers_.insert(std::make_pair(pointer_expr, array_expr));
+  auto it_bool =
```
Could be `const`.

LAJW reviewed (approved):
Looks alright, but I have some suggestions on how it could be better.

romainbrenguier committed:
Remove redundant pointer to array association

This redundant step which is already done in process_parameters, was
detected in string constraint generation as an attempt to access two
array to the same pointers.
Although the array are in fact the same here, the second call to
add_pointer_to_array_association is unecessary.

allredj commented:
PR on test-gen is here: https://github.com/diffblue/test-gen/pull/1231

tautschnig commented on src/java_bytecode/java_object_factory.cpp:
```diff
@@ -593,21 +611,38 @@ codet initialize_nondet_string_struct(
-    exprt data_expr = make_nondet_infinite_char_array(symbol_table, loc, code);
+    const dereference_exprt data_expr(data_pointer);
+    const exprt nondet_array =
+      make_nondet_infinite_char_array(symbol_table, loc, code);
```
This is somewhat redundant as leaving it uninitialised has the same effect?!

romainbrenguier replied:
From the point of view of verification yes, but in the trace it is better to have an explicit assignment so that we can see what value was given to the array.
